### PR TITLE
Integrate halogen and setup canvas drawing

### LINKED ISSUE #5: Integrate Halogen
## Details:
Hours Spent: 2

## Background
Halogen is one of the most common view frameworks for PS. It can easily have an HTML canvas integrated into it allowing the plot area to be implemented. [See Halogen](https://github.com/purescript-halogen/purescript-halogen)

## Overview
The first step to implement reach the goals of M1 is to have halogen integrated into the project so that the user can view a web page and the PS is bundled correctly.

## Acceptance Criteria
- [x] User can view a webpage that is constructed using halogen
- [x] Code is bundled into a single JS file

## Closing Notes

Overview of changes can be found in PR #11

### LINKED ISSUE #6: Set up canvas
## Details:
Hours Spent: 13

## Background
The canvas is how the user will view the plot. This canvas needs to be able to display all the features that are found in [`CanvasPlotter.hs`](https://github.com/michalkonecny/exact-function-plotter/blob/master/canvas/src-ghcjs/CanvasPlotter.hs)

## Overview
The canvas should be embedded in the halogen view. There should be a button added to the view that can be made to call any of the features like draw polygon. This is simply a proof of concept and hence the appearance of these features can be refactored later.

## Acceptance Criteria
- [X] There is a UI button that will draw a simple polygon on the canvas

## Closing Notes

Overview of changes can be found in #11 

It should be noted that there is currently no support for stopping plot computation in the middle of its execution as this requires a unique way to handle doing large computations.

### PR DESCRIPTION
**Pull request by RelativeForce**
# Issue(s)
closes #6 
closes #5 

# Overview of changes
- Integrated Halogen UI framework into the project
- Updated `npm run build` to build and bundle the app into `dist/app.js`
- Added `app.js` and `debug.log` to gitignore
- Fixed error in `README.md`
- Added `Components.Canvas` component for abstracting drawing on the JS canvas.
- Added `DrawCommand` monad as a DSL for the different drawing commands that are interpreted by the `runDrawCommand` function in `Components.Canvas.Commands.Interpreter`. The canvas is responsible for drawing the drawing commands but not computing them.
- Added the `Components.Canvas.CanvasController` module for handling setting up the canvas, rendering the `DrawCommand`s and handling if the canvas is resized.
- Added `Components.Canvas.Commands.Actions` to contain the functions that construct the `DrawCommand` DSL language commands.
- Ported the draw functions from the old version of the application into `Components.Canvas.Draw`
- Added `Plot` as the data interface between the UI and the actual drawing command computations.
- Added `PlotController` which is responsible for computing the `Plot` data into the `DrawCommand`s that will be performed on the canvas. This performs this task asynchronously.
- Added record types to describe the data that is used in computations.

# Unit test changes
N/A

### CODE REVIEW

File: src/Components/Canvas/Canvas.purs [hunk 1]
```diff
@@ -34,7 +34,7 @@ data Action operations
   = Init
   | HandleInput (Input operations)
 
-canvasComponent :: forall operations query output m. MonadEffect m => Renderer operations -> H.Component HH.HTML query (Input operations) output m
+canvasComponent :: forall operations query output m. MonadEffect m => CanvasController operations -> H.Component HH.HTML query (Input operations) output m
 canvasComponent cfg =
```
Comment (michalkonecny) [inline src/Components/Canvas/Canvas.purs]:
```suggestion
canvasComponent controller =
```
also renaming in the function body

File: src/Main.purs [hunk 2]
```diff
@@ -2,9 +2,93 @@ module Main where
...
+          [ HE.onClick \_ -> Just BasicPlot ]
+          [ HH.text "Plot polygon async" ]
+      , HH.slot _canvas 1 (canvasComponent canvasController) state.input absurd
+      ]
+
+computePlot :: Plot -> ReaderT Config Aff (DrawCommand Unit)
+computePlot plot = do
+  config <- ask
+  result <- lift $ computePlotAsync plot
+  pure $ result
```
Comment (michalkonecny) [inline src/Main.purs]: Currently this code is equivalent to `lift $ computePlotAsync plot`.  Is the extra there in preparation for something more complex?

**Review by michalkonecny — COMMENTED** [review]
I added a minor renaming suggestion in Canvas.purs and added a minor query to Main.purs.Overall I am very happy with it.

File: src/Main.purs (hunk 2 above)
Comment (RelativeForce) [reply src/Main.purs]: The `ask` allows us to access the config for the component which may be useful at some point but for the time being ill simplify it

File: src/Components/Canvas/Canvas.purs (hunk 1 above)
Comment (RelativeForce) [reply src/Components/Canvas/Canvas.purs]: Good spot

File: src/Main.purs (hunk 2 above)
Comment (RelativeForce) [reply src/Main.purs]: Resolved - Fixed in the latest commit

File: src/Components/Canvas/Canvas.purs (hunk 1 above)
Comment (RelativeForce) [reply src/Components/Canvas/Canvas.purs]: Resolved - Fixed in the latest commit

File: src/Main.purs (hunk 2 above)
Comment (michalkonecny) [reply src/Main.purs]: Thank you for the explanation and addressing it.  I did wonder whether the config was there ready for future use.  In future similar cases, I would be also happy leaving it there with a comment or commenting it out.

**Review by michalkonecny — APPROVED** [review]

**Pull request merged**